디자인 수정 및 Navigation 관련 프로토콜 프로퍼티 네이밍 수정

Hansolkkim opened the pull request:
1.   GoalAddVC, CategoryAddVC, GoalListCollectionViewCell에 <u>**Jee**</u>가 만들어주신 새로운 디자인 적용

     -   GoalAddVC

         |                        디자인 적용 전                        |                        디자인 적용 후                        |                         목표 디자인                          |
         | :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
         | ![SS2022-11-17PM02.52.30](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171452126.jpg) | <img src="https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171455090.jpg" alt="SS2022-11-17PM02.55.29" width="50%;" /><img src="https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171455714.jpg" alt="SS2022-11-17PM02.55.40" width="50%;" /> | ![SS2022-11-17PM02.54.27](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171454368.jpg) |

     -   CategoryAddVC

         |                        디자인 적용 전                        |                        디자인 적용 후                        |                         목표 디자인                          |
         | :----------------------------------------------------------: | :----------------------------------------------------------: | :----------------------------------------------------------: |
         | ![SS2022-11-17PM02.52.44](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171453760.jpg) | <img src="https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171456177.jpg" alt="SS2022-11-17PM02.56.06" width="50%;" /><img src="https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171456660.jpg" alt="SS2022-11-17PM02.56.12" width="50%;" /> | ![SS2022-11-17PM02.56.25](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171456434.jpg) |

     -   GoalListCollectionViewCell

         | 디자인 적용 전                                               | 디자인 적용 후                                               | 목표 디자인                                                  |
         | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
         | ![SS2022-11-17PM06.00.24](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171800240.jpg) | ![SS2022-11-17PM05.36.00](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171736480.jpg) | ![SS2022-11-17PM06.00.55](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171801713.jpg) |

     -   그 이외 추가적으로 GoalListVC의 rightBarButtonItem 색상 변경 및 TabBarItem의 사이즈 수정 등을 진행했습니다.
         <img src="https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171816737.jpg" alt="SS2022-11-17PM06.16.45" width="30%;" />

          예전에 Jee 께서 탭바 아이콘이 너무 크고, 초록색은 버리는게 좋을거 같다고 해서 일단 임의대로 수정했습니다..ㅎㅎ

1.   CategoryAddVC의 formView 사라지는 애니메이션 추가

     -   CategoryAddVC의 formView는 BottomSheet을 사용한 커스텀 UI가 아니여서 사라지는 애니메이션에 대한 로직이 별도로 필요했던거 같습니다.
     -   그래서 최근에 살펴봤던 **CGAffineTransform**을 이용하여 CategoryAddVC의 formView가 나타나고, 사라지는 기능을 구현했습니다.

1.   GoalAddVC, CategoryAddVC가 여러 디바이스에서 원하는 모습대로 보이도록 UI의 Constraints 수정

     -   Jee께서 주신 디자인대로 적용하려고 Constraint들을 모두 상수로 지정해주다보니 다른 디바이스에서는 원하는 모습으로 UI들이 보이지 않는 상황이 발생했습니다.
         (예시는 아이폰8 에서 확인한 화면입니다.)

         | GoalAddVC                                                    | CategoryAddVC                                                |
         | ------------------------------------------------------------ | ------------------------------------------------------------ |
         | ![SS2022-11-17PM03.25.07](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171525450.jpg) | ![SS2022-11-17PM03.24.59](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171525634.jpg) |

         -   GoalAddVC 같은 경우에는 UI가 깨진다기 보다는 간격, 글씨 크기, Button Height가 화면에 맞지 않는 크기로 보였습니다.

     -   Constraint들을 현재 화면에 맞게 `UIScreen.main.bounds`를 곱하여 사용하는 상수로 수정하여 문제를 해결했습니다.

     -   또한 CategoryAddVC의 formView의 경우에는 Height를 직접 설정해주지 않고, SubView들의 Constraint로 그 Height가 설정되도록 변경했습니다. 그리고 취소 버튼과 저장 버튼도 일정 spacing을 가지는 StackView에 넣어 디바이스 화면 크기에 잘 적용되도록 했습니다.

     -   아래와 같이 아이폰 8에서도, 아이폰 13 Pro Max에서도 적절하게 UI가 보이는 것을 확인했습니다.

         | GoalAddVC - 아이폰8                                          | CategoryAddVC - 아이폰8                                      | GoalAddVC - 아이폰13ProMax                                   | CategoryAddVC - 아이폰13ProMax                               |
         | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
         | ![SS2022-11-17PM03.31.06](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171531006.jpg) | ![SS2022-11-17PM03.31.12](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171531563.jpg) | ![SS2022-11-17PM03.31.51](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171531885.jpg) | ![SS2022-11-17PM03.31.45](https://raw.githubusercontent.com/Hansolkkim/Image-Upload/forUpload/img/202211171531809.jpg) |

         
1. 그리고 이전에 #98 에서 토의했던 내용대로, Navigation 관련 프로토콜들의 프로퍼티 네이밍을 수정했습니다!

Hansolkkim committed:
rename: [#98] Navigation 프로토콜 관련 프로퍼티 네이밍 수정

Hansolkkim committed:
design: [#104] GoalAddVC, CategoryAddVC에 새로운 디자인 적용

Hansolkkim committed:
feat: [#104] TextField 관련 수정

- 사용자 편의를 위하여 TextField 사용시 키보드 return 버튼 터치 시 키보드가 사라지도록 하는 기능 추가
- TextField에 추가적으로 구현되어 있던 자동 대문자, 오타 수정, 맞춤법 검사, smart dash 등의 기능 제거

Hansolkkim committed:
feat: [#104] CategoryAddVC의 FormSheet 사라지는 애니메이션 기능 추가

- CategoryAddVC의 FormSheet이 나타나는 로직을 CGAffineTransform을 이용하는 로직으로 수정
- CategoryAddVC의 FormSheet이 사라지는 로직 추가

Hansolkkim committed:
design: [#104] CategoryAddVC, GoalAddVC 의 디자인이 모든 디바이스에서 적절해보이도록 UI의 Constraints 수정

Hansolkkim committed:
feat: [#104] GoalListCollectionViewCell에 새롭게 적용할 CircularProgressBar 구현

Hansolkkim committed:
design: [#104] GoalListCollectionViewCell에 새로운 디자인 적용

Hansolkkim committed:
chore: Period의 기본 생성자에서 UTC를 기준으로 한 시간으로 start, end를 설정하도록 수정

Hansolkkim committed:
design: [#104] TabBarItem 사이즈 수정 및 그 외 자잘한 디자인 수정

- TabBarItem 아이콘 사이즈 수정
- TabBarItem 아이콘 뒤에 덧씌워지는 원 크기 및 색상 수정
- GoalListVC의 rightBarButtonItem 색상 수정

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/Views/InsertField.swift:
```diff
@@ -46,6 +53,18 @@ final class InsertField: UIControl {
+        }
+    }
+    
+    var textSize: CGFloat? {
+        didSet {
+            textField.font = UIFont.systemFont(ofSize: (textSize ?? 10)/750 * UIScreen.main.bounds.height)
+        }
+    }
```
InsertField의 TextField들의 폰트 크기도 작게 바꾸는게 작은 화면 디바이스(아이폰8)에서 더 이뻐보여서 해당 프로퍼티도 추가했습니다!

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/Views/InsertField.swift:
```diff
@@ -46,6 +53,18 @@ final class InsertField: UIControl {
         }
     }
     
+    var textColor: UIColor? {
+        didSet {
+            textField.textColor = textColor
+        }
+    }
```
이전에 회의때 말씀드린 내용처럼 GoalAddVC에서 Period, Time의 경우엔 초깃값이 있는 상태라 억지로 placeholder처럼 바꿔줄 필요가 있어 해당 프로퍼티도 추가했습니다.

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/GoalAddViewController.swift:
```diff
@@ -131,6 +144,14 @@ final class GoalAddViewController: UIViewController {
+                viewController.timeField.text = text
+                viewController.timeField.textColor = UIColor(hexRGB: "#A7A7A7")
+            }
+            .disposed(by: disposeBag)
```
여기서 이 코드들을 이용해 억지로 placeholder를 넣어줬습니다..ㅋㅋ 사용자가 해당 InsertField를 터치하기 전까지만 placeholder와 폰트 색깔이 유지되도록 했어요!

Hansolkkim commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -149,34 +178,58 @@ final class CategoryAddViewController: UIViewController {
+    }
+}
+
+private extension CategoryAddViewController {
+    func showFormView() {
+        UIView.animate(withDuration: 0.2) {
+            let y = -(self.view.frame.height + self.formView.frame.height)/2
+            self.formView.transform = CGAffineTransform(translationX: 0, y: y)
+        }
+    }
+    
+    func hideFormView() {
+        UIView.animate(withDuration: 0.2) {
+            self.formView.transform = .identity
+        } completion: { [weak self] _ in
+            self?.viewModel?.closeButtonTouched.accept(())
         }
```
CategoryAddVC의 경우, 따로 사라지는 애니메이션이 없어 CGAffineTransform을 이용해 사라지는 애니메이션을 추가해줬고, 나타나는 애니메이션도 통일성있게 CGAffineTransform을 이용하는게 보기 좋을 것 같아서 `showFormView()` 메소드도 수정했어요!

junu0516 replied:
오... 이건 처음들어보네요 ㅋㅋ
저도 좀 찾아보니까 복잡한 뷰의 프레임 계산 없이 애니메이션을 처리하는듯한데(맞나요..?), 
BottomSheet 저번에 개선했을 때 올라가고 내려가는 애니메이션 처리를 화면의 높이를 늘리고 줄이고가 아닌, y값을 변경시키는 정도로 해서 동일하게 적용해볼 수도 있을 것 같네요!

junu0516 replied:
아니 다시보니 애초에 목적지 좌표만 알면 되지 더 편하네요 ㅋㅋㅋ

Hansolkkim replied:
오 그렇긴한데, 그 아래로 스와이프해서 내리는 거는 적용하려면 또 굉장히 어렵지 않을까,,,생각이 듭니다ㅋㅋㅋㅋㅋ

Hansolkkim commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -149,34 +178,58 @@ final class CategoryAddViewController: UIViewController {
             $0.centerX.equalToSuperview()
             $0.top.equalTo(view.safeAreaLayoutGuide.snp.bottom)
-            $0.width.equalToSuperview().multipliedBy(0.7)
-            $0.height.equalToSuperview().multipliedBy(0.18)
```
BottomSheet을 사용하지 않아서 생긴 장점인데, 따로 FormView의 Height를 지정해주지 않고 subview들의 constraint로 FormView의 height가 유추가능하도록 했더니, subview들의 크기에 맞게 formView의 height가 알아서 맞춰졌어요!

junu0516 replied:
BottomSheet.... 뭔가 파고들수록 슅이 되는거같아욬ㅋㅋ 
저번에 공유창 띄우셨을 때 레이아웃 또 깨지는것도 그렇고... 점점 진지하게 라이브러리의 도입..을 해야되는걸까 생각이 듭니다 ㅋㅋㅋ

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/GoalAddViewController.swift:
```diff
@@ -120,6 +125,14 @@ final class GoalAddViewController: UIViewController {
             .bind(to: periodField.rx.text)
             .disposed(by: disposeBag)
+
+        Observable.just("목표 기간을 설정하세요.")
+            .withUnretained(self)
+            .bind { viewController, text in
+                viewController.periodField.text = text
+                viewController.periodField.textColor = UIColor(hexRGB: "#A7A7A7")
+            }
+            .disposed(by: disposeBag)
```
여기서 이 코드들을 이용해 억지로 placeholder를 넣어줬습니다..ㅋㅋ 사용자가 해당 InsertField를 터치하기 전까지만 placeholder와 폰트 색깔이 유지되도록 했어요!

junu0516 commented on rabit/rabit/Album/AlbumCoordinator.swift:
```diff
@@ -3,7 +3,7 @@ import RxSwift
 import RxRelay
 
 protocol AlbumNavigation {
-    var showPhotoEditView: PublishRelay<BehaviorRelay<Album.Item>> { get }
+    var didSelectPhoto: PublishRelay<BehaviorRelay<Album.Item>> { get }
     var didChangePhoto: PublishRelay<Void> { get }
 }
```
아.. 전에 제안해주셨던 네이밍 수정 몰아서 쫙 해야겠네요 ㅋㅋ..

junu0516 replied:
는 밑에보니 이미 다 해주셨군요 ㅋㅋㅋㅋ 진짜 감사합니다... 게으름 다시한번 더 사죄 ^_ㅠ

junu0516 commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -15,42 +15,71 @@ final class CategoryAddViewController: UIViewController {
         return view
     }()
+
+    private let titleLabel: UILabel = {
+        let label = UILabel()
+        label.text = "카테고리 생성"
+        label.font = UIFont.systemFont(
+            ofSize: 20/750 * UIScreen.main.bounds.height,
+            weight: .init(rawValue: 700)
+        )
+        return label
+    }()
```
12미니에서 확인했을 때도 의도한대로 잘 나타납니당 ㅎㅎ

junu0516 commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -15,42 +15,71 @@ final class CategoryAddViewController: UIViewController {
-        textField.leftView = UIView()
+        textField.placeholder = "카테고리를 입력하세요."
+        textField.autocapitalizationType = .none
+        textField.spellCheckingType = .no
+        textField.smartDashesType = .no
+        textField.autocorrectionType = .no
         textField.layer.borderColor = UIColor.systemGray4.cgColor
         textField.layer.borderWidth = 1.0
+        textField.font = UIFont.systemFont(ofSize: 15/750 * UIScreen.main.bounds.height)
         textField.roundCorners(10)
+        textField.addLeftPadding(width: 20)
         return textField
     }()
     
+    private var buttonStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.axis = .horizontal
+        stackView.distribution = .fillEqually
+        stackView.spacing = 10/376 * UIScreen.main.bounds.width
+        return stackView
+    }()
+    
     private lazy var closeButton: UIButton = {
         let button = UIButton()
         
-        button.setTitle("\t닫기\t", for: .normal)
+        button.setTitle("취소", for: .normal)
```
원래는 패딩 바로 줬어야 된건데 ㅋㅋㅋㅋ.. 게을러서 \t로 가라친거 다시 보니 부끄럽네요

junu0516 commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -86,20 +115,11 @@ final class CategoryAddViewController: UIViewController {
-            animations: self.view.layoutIfNeeded,
-            completion: nil
-        )        
+        saveButton.roundCorners(saveButton.frame.height / 2)
+        closeButton.roundCorners(closeButton.frame.height / 2)
```
이건 필수는 아닌데 동적으로 레이아웃값 가져와서 조작하는거 함수 하나로 몰아도 괜찮을듯해요!

Hansolkkim replied:
엇 이거 무슨 말씀인지 다시 한 번만 설명 부탁드려요! `viewDidLayoutSubviews()` 에서 호출할 메소드로 하나 빼자는 말씀이신가요??

junu0516 commented on rabit/rabit/Goal/CategoryAdd/CategoryAddViewController.swift:
```diff
@@ -136,6 +158,13 @@ final class CategoryAddViewController: UIViewController {
         viewModel.warningLabelHidden
             .bind(to: warningLabel.rx.isHidden)
             .disposed(by: disposeBag)
+        
+        textField.rx.controlEvent(.editingDidEndOnExit)
+            .withUnretained(self)
+            .bind { viewController, _ in
+                viewController.textField.resignFirstResponder()
+            }
+            .disposed(by: disposeBag)
```
이거 혹시 전에 언급해주셨던 문제들하고 관련있는건가요?? (정신이 없어서 제가 까먹은거 같아요 ㅠㅠ)
텍스트필드에 뭔가를 입력하고 끝낼 때 responder chain에서 제일 먼저 반응하던거 취소하는거 같은데.. 혹시 동작의 의도 한번만 더 여쭤봐도 될까요!!?

Hansolkkim replied:
아아 이거 키보드에서 retrun 버튼 눌렀을 때 키보드 내려가는 기능 추가한거에요!
키보드에서 return 버튼 눌린거를 `.editingDidEndOnExit` 으로 받더라구요. 그리고 textField가 `firstResponder`가 되는 때는 키보드가 올라온 때라 이걸 취소하는거였습니다ㅎㅎ

junu0516 commented on rabit/rabit/Goal/Views/GoalListCollectionViewCell.swift:
```diff
@@ -14,16 +14,23 @@ final class GoalListCollectionViewCell: UICollectionViewCell {
+        return stackView
+    }()
+    
+    private let goalProgressView: CircularProgressBar = {
+        let progressView = CircularProgressBar()
+        progressView.backgroundColor = .clear
+        return progressView
     }()
```
이거 구현해주신 것도 봤는데 혹시 UI 확인 아직 안되는거죠!??
코드 돌려보는 중인데 아직 실제 적용전인가 싶어서요!!

Hansolkkim replied:
Goal 모델의 target, progress 값을 적절하게 잘 조절해주시면 확인이 됩니다!
progress의 경우 Realm에서 읽어오는 때에 사진의 개수에 맞게 수정하는 부분이 있어서 아마 임의로 만들고 사진 추가 안한 Goal의 경우에는 UI 확인이 안될거에요!
그리고 target이 0일 경우에도 UI 확인이 안됩니다ㅠㅠ

junu0516 reviewed (commented):
수고 많으셨어요 ㅋㅋㅋ 디자인 훨씬 예뻐졌네요..!
새로 수정할건 없고 제가 처음 배운거랑 질문할 것만 남겼어요 ㅋㅋㅋ
커밋 편하게 추가하시고 더 이상 없으심 머지하시면 될듯합니다~!
토요일에 봬요!!

Hansolkkim committed:
design: [#104] InsertField에서 사용할 아이콘 추가 및 코드 변경

Hansolkkim committed:
design: [#104] GoalDetailVC의 UI들도 디바이스 크기에 맞게 변경되도록 수정

Hansolkkim committed:
chore: ImageSet 변경 적용

Hansolkkim committed:
chore: GoalListVC의 GoalList를 리로드하는 타이밍을 viewWillAppear로 수정

- 기존에는 viewDidLoad에서 리로드하도록 했는데, 이 경우 사진을 찍은 후 GoalAddVC를 pop 하더라도 해당 Cell의 ProgressBar가 업데이트되지 않는 문제가 발생하여, 해당 로직을 viewWillAppear에서 처리하도록 수정

Hansolkkim committed:
chore: [#104] ProgressBar 상하좌우 깎이는 문제 해결

- 아이폰11에서 ProgressBar의 상하좌우가 깎이는 문제가 있어, ProgressBar의 반지름 수정
